porch/api/porchconfig/v1alpha1/types_functions.go

```diff
@@ -0,0 +1,107 @@
+// to ensure uniqueness of names, and will follow formatting of
+// [DNS Subdomain Names](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).
+type Function struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   FunctionSpec   `json:"spec,omitempty"`
+	Status FunctionStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// FunctionList
+type FunctionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []Function `json:"items"`
+}
+
+type FunctionType string
+
+const (
+	FunctionTypeValidator FunctionType = "validator"
+	FunctionTypeMutator   FunctionType = "mutator"
+)
+
+// FunctionSpec defines the desired state of a Function
+type FunctionSpec struct {
+	// Image specifies the function image, such as 'gcr.io/kpt-fn/gatekeeper:v0.2'.
+	Image string `json:"image"`
+
+	// RepositoryRef references the repository in which the function is located.
+	RepositoryRef RepositoryRef `json:"repositoryRef"`
+
+	// FunctionType specifies the function types (mutator, validator or/and others).
+	FunctionTypes []FunctionType `json:"functionTypes,omitempty"`
+
+	FunctionConfigs []FunctionConfig `json:"functionConfigs,omitempty"`
+
+	// Keywords are used as filters to provide correlation in function discovery.
+	Keywords []string `json:"keywords,omitempty"`
+
+	// Description is a short description of the function.
+	Description string `json:"description"`
+
+	// `DocumentationUrl specifies the URL of comprehensive function documentation`
+	DocumentationUrl string `json:"documentationUrl,omitempty"`
+
+	// InputTypes specifies to which input KRM types the function applies. Specified as Group Version Kind.
+	// For example:
+	//
+	//    inputTypes:
+	//    - kind: RoleBinding
+	//      # If version is unspecified, applies to all versions
+	//      apiVersion: rbac.authorization.k8s.io
+	//    - kind: ClusterRoleBinding
+	//      apiVersion: rbac.authorization.k8s.io/v1
+	// InputTypes []metav1.TypeMeta
+
+	// OutputTypes specifies types of any KRM resources the function creates
+	// For example:
+	//
+	//     outputTypes:
+	//     - kind: ConfigMap
+	//       apiVersion: v1
+	// OutputTypes []metav1.TypeMeta
+
+}
+
+// FunctionConfig specifies all the valid types of the function config for this function.
+// If unspecified, defaults to v1/ConfigMap. For example, function `set-namespace` accepts both `ConfigMap` and `SetNamespace`
+type FunctionConfig struct {
+	metav1.TypeMeta `json:",inline"`
+	// Experimental: requiredFields tells necessary fields and is aimed to help users write the FunctionConfig.
+	// Otherwise, users can get the required fields info from the function evaluation error message.
+	RequiredFields []string `json:"requiredFields,omitempty"`
+}
+
+// FunctionStatus defines the observed state of Function
+type FunctionStatus struct {
```

Comment:
This can *eventually* provide info such as `stable` version of `latest` etc.

Reply:
Agree.  Right now this is a direct copy of the aggregated apiserver type, so we can migrate without worrying about anything more than the group changing, I hope!